proposals/jsr-305-custom-nullability-qualifiers.md

```diff
@@ -0,0 +1,321 @@
+
+```java
+@Nonnull
+@TypeQualifierDefault({ElementType.METHOD, ElementType.PARAMETER})
+public @interface NonNullApi {
+}
+
+@Nonnull(when = When.MAYBE)
+@TypeQualifierDefault({ElementType.METHOD, ElementType.PARAMETER})
+public @interface NullableApi {
+}
+```
+```java
+// FILE: test/package-info.java
+
+@NonNullApi // declaring all types in package 'test' as non-nullable by default
+package test;
+
+
+// FILE: test/A.java
+package test;
+
+@NullableApi // overriding default nullability from the package
+interface A {
+    String foo(String x); // fun foo(x: String?): String?
+ 
+    @NotNullApi // overriding default from the class
+    String bar(String x, @Nullable String y); // fun bar(x: String, y: String?): String 
+}
+
+```
+
+### `@Migration` annotation
+Current proposal suggests to add the following meta-annotation in 
+the `kotlin.annotation` package:
+```kotlin
+enum class MigrationStatus {
+    ERROR,
+    WARNING,
+    IGNORE
+}
+
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+annotation class Migration(val status: MigrationStatus)
+```
+
+Its aim is to define a migration status of an annotation it's been applied to.
+To the moment it can be only used for custom nullability qualifier 
+[nicknames](#Type qualifier nicknames) introduced in this proposal as well.
+
+When the annotation is applied to a nullability nickname its argument specifies
+how the compiler handles the nickname:
+- `MigrationStatus.ERROR` makes annotation work just the same way as any plain
+nullability annotation, i.e. reporting errors for inappropriate usages of an 
+annotated type. In other words, it's equivalent to absence of the 
+`@Migration` annotation.
+- `MigrationStatus.WARNING` should work just the same as 
+`MigrationStatus.ERROR`, but compilation warnings must be reported instead of 
+errors 
+for 
+inappropriate usages of an annotated type 
+- `MigrationStatus.IGNORE` makes compiler to ignore the annotation completely
+
+
+Inappropriate usages in Kotlin that are subjects to report errors/warnings
+include:
+- Passing nullable arguments for parameters those types are enhanced from
+annotations to not null
+- Using enhanced nullable values as an argument for kotlin not-nullable 
+parameter or as a receiver for a method call
+- Overriding enhanced Java methods with incorrect signature in Kotlin
+
+```java
+@TypeQualifierNickname
+@Nonnull(when = When.ALWAYS)
+@Migration(status = MigrationStatus.WARNING)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface MyNonnull {
+}
+
+
+interface A {
+    void foo(@MyNonnull String x);
+}
+
+```
+
+```kotlin
+fun bar(a: A) {
+    a.foo(null) // a warning will be reported on 'null' argument
+}
+```
+
+In the example the warning will be reported until status argument is changed
+to `ERROR`, then it should become an error.
+
+It's important that migration annotation works only for its immediate usages
+on types or through `TypeQualifierDefault` annotation.
```

Comment:
It says above
> it can be only used for custom nullability qualifier nicknames
, so this line is a bit confusing. Can I use it with `TypeQualifierDefault` or not? Maybe provide an example